Improve usability for Diagnostic Detail panel

defunctzombie opened the pull request:
**User-Facing Changes**
The Autocomplete widget is replaced with a different autocomplete widget. The overall experience is similar to before (you can type to filter the list). The new widget resolves some UX strangeness from the previous widget described in more detail below.

We've also removed the code paths which strip away a leading "/" from the hardware id as well as the code paths which remove the hardware_id string from the _name_ field. Both of these looked like webviz "cruisisms" and don't seem to be something that should be done for all ROS and non-ros users.

**Description**
The autocomplete in the Diagnostic Detail panel would hide available options once the user selected one. To change to a different diagnostic the user has to click the input, realize they can delete their text, delete all the text and type again to find their entry.

This change switches the autocomplete input to the mui Autocomplete component. This component exhibits better UX behavior. When focusing on the input, the user is shown all available entries with their current entry highlighted. This also resolves some glitches with the dropdown.

This change also modifies the autocomplete options logic to ignore empty hardware ids with no name from showing in the autocomplete options list. These entries would create blanks in the dropdown list.

Fixes: #2379



<!-- link relevant github issues -->
<!-- add `docs` label if this PR requires documentation updates -->

defunctzombie commented:
@jhurliman can you give this PR a spin and see how it feels to use especially for the issues raised in #2379 . Once we are happy with the interactions we can get feedback from others on our changes.

jhurliman commented:
@defunctzombie yes, this is better. Ship it

jhurliman reviewed (changes requested):
I think the stories `"DiagnosticStatusPanel:Selected Hardware ID Only"`, `"DiagnosticStatusPanel:Selected Name"`, and `"DiagnosticStatusPanel:Moved Divider"` should be updated. It sets the `name` field to values such as `hardware_id1: name1` which we assume is a Cruise-ism, so the output looks off with the redundant hardware ids.

Another thing that appears to have changed is the dropdown is always visible, even when there are zero or one hardware ids. I prefer the previous behavior of hiding the dropdown in those cases (I generally lean toward minimal UI for Studio panels), but it's a good discussion point.

defunctzombie commented:
> I think the stories `"DiagnosticStatusPanel:Selected Hardware ID Only"`, `"DiagnosticStatusPanel:Selected Name"`, and `"DiagnosticStatusPanel:Moved Divider"` should be updated. It sets the `name` field to values such as `hardware_id1: name1` which we assume is a Cruise-ism, so the output looks off with the redundant hardware ids.

I looked at this and it's actually not because the story sets the name field to `hardware_id1: name1` but because the `<DiagnosticStatus>` component which renders each DiagnosticInfo item renders the displayName field from each item. We could detect that you've filtered down to a specific hardware_id or even a hardware_id and name and pass a prop to the DiagnositcStatus component to avoid showing the heading but its starting to feel like a bit too much scope creep for me on this specific fix.

> Another thing that appears to have changed is the dropdown is always visible, even when there are zero or one hardware ids. I prefer the previous behavior of hiding the dropdown in those cases (I generally lean toward minimal UI for Studio panels), but it's a good discussion point.

What if there are multiple names under a single hardware id? It would be weird UX to me for the panel to automatically select a diagnostic when you open a datasource just because that happened to be the first message seen in the data source. I actually find this behavior for the image panel somewhat annoying and another one of those areas where the app is trying to be too clever instead of making it easy and obvious how to configure the panel with your data.

defunctzombie commented:
@jhurliman I had accidentally removed a floating from the panel toolbar when I was iterating. I've put that back.

jtbandes commented:
This text should probably be disabled (grayed out), or just hide the dropdown when waiting for diagnostics because we already say it in the center of the panel:

<img width="259" alt="image" src="https://user-images.githubusercontent.com/14237/150231152-f6cb070c-c84a-46e0-b68c-11c38e5bbd21.png">

We also need a drop shadow on the menu so it doesn't blend in with other UI elements:
<img width="269" alt="image" src="https://user-images.githubusercontent.com/14237/150231262-25be27a0-d40b-45b8-9a75-59aa05a7a1de.png">

compare with the Message Path Input autocomplete which has a shadow:
<img width="335" alt="image" src="https://user-images.githubusercontent.com/14237/150231250-3082cf76-0c77-4e79-ac86-c1ed897b4e34.png">

jtbandes reviewed (changes requested):
needs UI cleanup

defunctzombie commented:
I've updated the PR to disable the dropdown when there are no diagnostics - this prevents "waiting for diagnostics..." from being shown in the empty state.

I've updated the PR to set the default paper elevation to `2` which brings back a dropshadow.

@2metres Thoughts on why the default paper elevation was set to 0?

defunctzombie commented on packages/studio-base/src/theme/muiComponents.ts:
```diff
@@ -109,7 +109,7 @@ export default function muiComponents(theme: Theme): ThemeOptions["components"]
     MuiPaper: {
       defaultProps: {
-        elevation: 0,
+        elevation: 2,
```
@2metres is this ok to do here?

2metres replied:
Yep looks good to me.

defunctzombie commented:
@jtbandes take a look and if you are 👍 on the changes please squash and merge.
